presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java

```diff
@@ -64,6 +68,8 @@ public SubfieldExtractor(StandardFunctionResolution functionResolution)
                 RowType baseType = (RowType) base.getType();
 
                 RowExpression indexExpression = dereferenceExpression.getArguments().get(1);
+                indexExpression = expressionOptimizer.apply(indexExpression);
```

Comment:
combine this statement with the one above: `RowExpression indexExpression = expressionOptimizer .apply(dereferenceExpression.getArguments().get(1));`